### PR TITLE
[Fix] Fix bugs in `analyze_logs`

### PR DESCRIPTION
**Pull request by Xiangxu-0103**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Usage:

If users want to plot training parameters, take `lr` as an example:

```shell
python tools/analysis_tools/analyze_logs.py plot_curve ${JSON_FILE} --keys lr
```

If users want to plot evaluation metric, take `Overall_3D_AP11_moderate` as an example:

```shell
python tools/analysis_tools/analyze_logs.py plot_curve ${JSON_FILE} --keys Overall_3D_AP11_moderate --eval --eval-interval ${EVAL_INTERVAL}
```

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: tools/analysis_tools/analyze_logs.py [hunk 1]
```diff
@@ -50,55 +52,22 @@ def plot_curve(log_dicts, args):
         epochs = list(log_dict.keys())
         for j, metric in enumerate(metrics):
             print(f'plot curve of {args.json_logs[i]}, metric is {metric}')
-            if metric not in log_dict[epochs[args.interval - 1]]:
+            if metric not in log_dict[epochs[int(args.eval_interval) - 1]]:
                 raise KeyError(
-                    f'{args.json_logs[i]} does not contain metric {metric}')
-
-            if args.mode == 'eval':
```
Comment (JingweiZhang12) [inline tools/analysis_tools/analyze_logs.py:57]: Why the content of mode `eval` has been deleted?

File: tools/analysis_tools/analyze_logs.py [hunk 2]
```diff
@@ -17,18 +17,20 @@ def cal_train_time(log_dicts, args):
                 all_times.append(log_dict[epoch]['time'])
             else:
                 all_times.append(log_dict[epoch]['time'][1:])
-        all_times = np.array(all_times)
-        epoch_ave_time = all_times.mean(-1)
+        if not all_times:
+            raise KeyError(
+                'Please reduce the log interval in the config so that '
+                'interval is less than iterations of one epoch.')
+        epoch_ave_time = np.array(list(map(lambda x: np.mean(x), all_times)))
```
Comment (JingweiZhang12) [inline tools/analysis_tools/analyze_logs.py:24]: This modification which uses iteration may be slower than the prior code?

File: tools/analysis_tools/analyze_logs.py [hunk 3]
```diff
@@ -124,6 +93,16 @@ def add_plot_parser(subparsers):
         nargs='+',
         default=['mAP_0.25'],
         help='the metric that you want to plot')
+    parser_plt.add_argument(
+        '--start-epoch',
+        type=str,
+        default='1',
+        help='the epoch that you want to start')
+    parser_plt.add_argument(
+        '--eval-interval',
```
Comment (JingweiZhang12) [inline tools/analysis_tools/analyze_logs.py:102]: are you sure the added arguments `start-epoch` and `eval-interval` are used？

File: tools/analysis_tools/analyze_logs.py [hunk 4]
```diff
@@ -174,17 +151,28 @@ def load_json_logs(json_logs):
...
+
                 if epoch not in log_dict:
                     log_dict[epoch] = defaultdict(list)
+
                 for k, v in log.items():
-                    log_dict[epoch][k].append(v)
+                    if '/' in k:
+                        log_dict[epoch][k.split('/')[-1]].append(v)
+                        val_flag = True
+                    elif val_flag:
```
Comment (JingweiZhang12) [inline tools/analysis_tools/analyze_logs.py:168]: Are you sure about this line of modification? We'll appreciate it if you can run these `analyze_logs` scripts to make it bug-free.

Comment (Xiangxu-0103) [reply tools/analysis_tools/analyze_logs.py:168]: This script is just modified from [mmdet](https://github.com/open-mmlab/mmdetection/blob/dev-3.x/tools/analysis_tools/analyze_logs.py). I will check it ASAP.

**Review by JingweiZhang12 — COMMENTED** [review]
@Xiangxu-0103 Hi, please update this PR if you have checked the correctness of running.

File: tools/analysis_tools/analyze_logs.py (hunk 1 above)
Comment (Xiangxu-0103) [reply tools/analysis_tools/analyze_logs.py:57]: Original implementation will raise a `TypeError`.

File: tools/analysis_tools/analyze_logs.py (hunk 2 above)
Comment (Xiangxu-0103) [reply tools/analysis_tools/analyze_logs.py:24]: Original implementation will raise a `TypeError`.

File: tools/analysis_tools/analyze_logs.py (hunk 3 above)
Comment (Xiangxu-0103) [reply tools/analysis_tools/analyze_logs.py:119]: Removed unused `start-epoch`, and `eval-interval` is used to plot evaluation metric.

**Comment by Xiangxu-0103** [comment]
> @Xiangxu-0103 Hi, please update this PR if you have checked the correctness of running.

Have checked this script.

File: tools/analysis_tools/analyze_logs.py (hunk 1 above)
Comment (Xiangxu-0103) [reply tools/analysis_tools/analyze_logs.py:57]: The `json` file does not content `mode` key. If users want to plot evaluation metric, we can set `args.eval` to True.

**Comment by sunjiahao1999** [comment]
LGTM.

**Review by sunjiahao1999 — APPROVED** [review]

File: tools/analysis_tools/analyze_logs.py [hunk 5]
```diff
@@ -50,56 +52,41 @@ def plot_curve(log_dicts, args):
...
             print(f'plot curve of {args.json_logs[i]}, metric is {metric}')
-            if metric not in log_dict[epochs[args.interval - 1]]:
+            if metric not in log_dict[epochs[int(args.eval_interval) - 1]]:
+                if args.eval:
+                    raise KeyError(
+                        f'{args.json_logs[i]} does not contain metric '
+                        f'{metric}. Please check if "--no-validate" is '
+                        'specified when you trained the model. Or check '
+                        f'if the eval_interval {args.eval_interval} in args '
+                        'is_equal to the eval_interval during training.')
```
Comment (ZwwWayne) [inline tools/analysis_tools/analyze_logs.py:62]: is_equal -> is equal to the ``eval_interval``

**Comment by WuYanXingege** [comment]
> @Xiangxu-0103 Hi, please update this PR if you have checked the correctness of running.

when i run python tools/analysis_tools/analyze_logs.py plot_curve 79.7\ 496\ 432/20230902_001556/vis_data/20230902_001556.json  --keys loss --eval
i encounter :ValueError: x and y must have same first dimension, but have shapes (399,) and (400,)
whats this error?

File: tools/analysis_tools/analyze_logs.py (hunk 1 above)
Comment (WuYanXingege) [reply tools/analysis_tools/analyze_logs.py:57]: when i run python tools/analysis_tools/analyze_logs.py plot_curve 79.7\ 496\ 432/20230902_001556/vis_data/20230902_001556.json --keys loss --eval
i encounter :ValueError: x and y must have same first dimension, but have shapes (399,) and (400,)
whats this error?

**Pull request merged**